src/array.jl

```diff
@@ -804,39 +804,57 @@ function levels!(A::CategoricalArray{T, N, R}, newlevels::Vector;
     A
 end
 
-function _unique(::Type{S},
-                 refs::AbstractArray{T},
-                 pool::CategoricalPool) where {S, T<:Integer}
-    nlevels = length(levels(pool)) + 1
-    order = fill(0, nlevels) # 0 indicates not seen
-    # If we don't track missings, short-circuit even if none has been seen
-    count = S >: Missing ? 0 : 1
-    @inbounds for i in refs
-        if order[i + 1] == 0
-            count += 1
-            order[i + 1] = count
-            count == nlevels && break
+# return unique refs (each value is unique) in the order of appearance in `refs`
+# similar to fallback Base.unique() implementation,
+# but short-circuits once references to all levels are encountered
+function _uniquerefs(A::AbstractCategoricalArray{T}) where T
+    arefs = refs(A)
+    res = similar(arefs, 0)
+    nlevels = length(levels(A))
+    seen = fill(false, nlevels+1) # +1 for 0 (missing ref)
+    maxunique = nlevels + (T >: Missing ? 1 : 0)
+    @inbounds for ref in arefs
+        if !seen[ref + 1]
+            push!(res, ref)
+            seen[ref + 1] = true
+            (length(res) == maxunique) && break
         end
     end
-    S[i == 1 ? missing : levels(pool)[i - 1] for i in sortperm(order) if order[i] != 0]
+    return res
 end
 
 """
-    unique(A::CategoricalArray)
+    uniquelevels(A::AbstractCategoricalArray)
```

Comment:
`unique=` keyword would be a bit obscure, since the levels are unique by definition. It could be something like `byoccurrence=` or `byfirstuse=`.

But I'm not sure this functionality is really needed. It would be nice to keep the returned value of `levels(catarr)` as it is -- independent of the actual stored data and identical to the result of `levels(pool(catarr))` call. And if one needs the original `unique(catarr)`, it would be just `unwrap.(unique(catarr))`. So maybe we just remove `uniquelevels()` and keep only the new `unique()`.

Reply:
Yeah, let's not add it for now and see whether `unwrap.(unique(x))` is enough.